File: library/src/main/java/cn/wandersnail/http/upload/ProgressRequestBody.java
package cn.wandersnail.http.upload;

import androidx.annotation.NonNull;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

import cn.wandersnail.http.util.HttpUtils;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.BufferedSink;

/**
 * date: 2019/8/23 18:17
 * author: zengfansheng
 */
class ProgressRequestBody extends RequestBody {
    private final MediaType contentType;
    private final FileInfo fileInfo;
    private final InternalUploadListener listener;

    ProgressRequestBody(MediaType contentType, FileInfo fileInfo, @NonNull InternalUploadListener listener) {
        this.contentType = contentType;
        this.fileInfo = fileInfo;
        this.listener = listener;
    }

    @Override
    public MediaType contentType() {
        return contentType;
    }

    @Override
    public long contentLength() throws IOException {
        return fileInfo.getFile() == null ? fileInfo.getInputStream().available() : fileInfo.getFile().length();
    }

    @Override
    public void writeTo(@NonNull BufferedSink sink) throws IOException {
        long uploadCount = 0;
        InputStream inputStream = null;
        try {
            byte[] buffer = new byte[10240];
            int len;
            if (fileInfo.getFile() != null) {
                inputStream = new FileInputStream(fileInfo.getFile());
            } else {
                inputStream = fileInfo.getInputStream();
            }
            while ((len = inputStream.read(buffer)) != -1) {
                sink.write(buffer, 0, len);
                uploadCount += len;
                listener.onProgress(fileInfo, uploadCount, contentLength());
            }
            listener.onComplete(fileInfo);
        } finally {
            HttpUtils.closeQuietly(inputStream);
        }
    }
}

File: library/src/main/java/com/snail/network/upload/ProgressRequestBody.kt
package com.snail.network.upload

import com.snail.network.callback.ProgressListener
import okhttp3.MediaType
import okhttp3.RequestBody
import okhttp3.internal.Util
import okio.BufferedSink
import okio.Okio
import okio.Source
import java.io.File
import java.io.IOException

/**
 *
 *
 * date: 2019/2/28 12:25
 * author: zengfansheng
 */
internal class ProgressRequestBody(private val contentType: MediaType?, private val file: File, private val listener: ProgressListener?) : RequestBody() {
    override fun contentType(): MediaType? {
        return contentType
    }

    override fun contentLength(): Long {
        return file.length()
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        var source: Source? = null
        var uploaded = 0L
        val uploadCount = 50000L//一次上传多少
        try {
            source = Okio.source(file)
            source.read(sink.buffer(), uploadCount)
            uploaded += uploadCount
            if (contentLength() > 0 && uploaded > 0) {
                listener?.onProgress(uploaded, contentLength())
            }
        } finally {
            Util.closeQuietly(source)
        }
    }
}